corim/unsignedcorim.go

```diff
@@ -4,104 +4,219 @@
+// tags array of the unsigned-corim-map
+func (o *UnsignedCorim) AddComid(c comid.Comid) *UnsignedCorim {
+	if o != nil {
+		if c.Valid() != nil {
+			return nil
+		}
+
+		comidCBOR, err := c.ToCBOR()
+		if err != nil {
+			return nil
+		}
 
-	var u comid.UUID
+		// d9 01fa # tag(506)
+		comidTag := []byte{0xd9, 0x01, 0xfa}
 
-	if dm.Unmarshal(data, &u) == nil {
-		o.val = u
-		return nil
-	}
+		taggedComid := append(comidTag, comidCBOR...)
 
-	return fmt.Errorf("unknown corim-id type (CBOR: %x)", data)
+		o.Tags = append(o.Tags, taggedComid)
+	}
+	return o
 }
 
-type Tag []byte
+// AddCoswid appends the CBOR encoded (and appropriately tagged) CoSWID to the
+// tags array of the unsigned-corim-map
+func (o *UnsignedCorim) AddCoswid(c swid.SoftwareIdentity) *UnsignedCorim {
+	if o != nil {
+		// Currently the swid package doesn't offer an interface
+		// for validating the supplied CoSWID, so -- for now --
+		// we take any input for granted and pass it to the encoder.
+		// See also https://github.com/veraison/swid/issues/23.
 
-type TaggedComid comid.Comid
+		coswidCBOR, err := c.ToCBOR()
+		if err != nil {
+			return nil
+		}
 
-type Locator struct {
-	Href       comid.TaggedURI `cbor:"0,keyasint" json:"href"`
-	Thumbprint swid.HashEntry  `cbor:"1,keyasint,omitempty" json:"thumbprint,omitempty"`
-}
+		// d9 01f9 # tag(505)
+		coswidTag := []byte{0xd9, 0x01, 0xf9}
 
-func NewUnsignedCorim() *UnsignedCorim {
-	return &UnsignedCorim{}
+		taggedCoswid := append(coswidTag, coswidCBOR...)
+
+		o.Tags = append(o.Tags, taggedCoswid)
+	}
+	return o
 }
 
-func (o *UnsignedCorim) SetIDString(s string) *UnsignedCorim {
+// AddDependentRim creates a corim-locator-map from the supplied arguments and
+// appends it to the dependent RIMs in the unsigned-corim-map
+func (o *UnsignedCorim) AddDependentRim(href string, thumbprint *swid.HashEntry) *UnsignedCorim {
 	if o != nil {
-		if s == "" {
-			return nil
+		l := Locator{
+			Href:       comid.TaggedURI(href),
+			Thumbprint: thumbprint,
+		}
+
+		if o.DependentRims == nil {
+			o.DependentRims = new([]Locator)
 		}
-		o.ID.val = s
+
+		*o.DependentRims = append(*o.DependentRims, l)
+
 	}
 	return o
 }
 
-func (o *UnsignedCorim) SetIDUUID(u comid.UUID) *UnsignedCorim {
+// AddProfile appends the supplied profile identifier (either a URL or OID) to
+// the profiles array in the unsigned-corim-map
+func (o *UnsignedCorim) AddProfile(urlOrOID string) *UnsignedCorim {
 	if o != nil {
-		if u == (comid.UUID{}) {
+		p, err := eat.NewProfile(urlOrOID)
+		if err != nil {
 			return nil
 		}
-		o.ID.val = u
+
+		if o.Profiles == nil {
+			o.Profiles = new([]eat.Profile)
+		}
+
+		*o.Profiles = append(*o.Profiles, *p)
 	}
 	return o
 }
 
-func (o UnsignedCorim) GetIDString() (string, error) {
-	switch t := o.ID.val.(type) {
-	case string:
-		return t, nil
-	default:
-		return "", fmt.Errorf("corim-id type is: %T", t)
+func (o UnsignedCorim) Valid() error {
+	if o.ID == (swid.TagID{}) {
```

Comment:
I am thinking as a future enhancement, should we move TagID as a common package to be used both by CoSWID/SWID and CoMID packages ? Here checking o.ID == swid.TagID{} looks slightly odd to me ??

Reply:
I don't understand what is the oddity here: the check makes sure the tag is not empty.

Reply:
kindly check my oddity from https://github.com/veraison/corim/issues/9  perspective, what i am trying to say !

If you agree, we can resolve this conversation.

Reply:
unfortunately I can't understand that either :-)